lib/Transforms/Instrumentation/GeneratePartialFnPass.cc

```diff
@@ -0,0 +1,118 @@
+
+#include "seahorn/Transforms/Instrumentation/GeneratePartialFnPass.hh"
+
+#include "seahorn/Analysis/SeaBuiltinsInfo.hh"
+#include "seahorn/InitializePasses.hh"
+#include "seahorn/Passes.hh"
+#include "seahorn/Support/SeaDebug.h"
+
+#include "llvm_seahorn/IR/LLVMContext.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+
+#define PARTIAL_FN_ANNOTATION "partial"
+
+using namespace llvm;
+
+namespace seahorn {
+
+const std::string GeneratePartialFnPass::PREFIX("sea.synth.inferable.");
+
+GeneratePartialFnPass::GeneratePartialFnPass() : ModulePass(ID) {}
+
+void GeneratePartialFnPass::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.addRequired<SeaBuiltinsInfoWrapperPass>();
+}
+
+bool GeneratePartialFnPass::runOnModule(Module &M) {
+  auto &SBI = getAnalysis<SeaBuiltinsInfoWrapperPass>().getSBI();
+  m_assumeFn = SBI.mkSeaBuiltinFn(SeaBuiltinsOp::ASSUME, M);
+
+  bool changed = false;
+  for (auto &F : M) {
+    if (!F.empty()) {
+      changed |= runOnFunction(SBI, F);
+    }
+  }
+  return changed;
+}
+
+bool GeneratePartialFnPass::runOnFunction(SeaBuiltinsInfo &SBI, Function &F) {
+  // Provides a placeholder implementation to each partial function stub.
+  bool changed = false;
+  for (auto &B : F) {
+    for (auto &I : B) {
+      IRBuilder<> builder(&B);
+
+      // Does I call a user defined function.
+      CallBase *CI = dyn_cast<CallBase>(&I);
+      if (!CI || SBI.isSeaBuiltin(*CI))
+        continue;
+
+      // Is the callee defined.
+      Function *CF = CI->getCalledFunction();
+      if (!CF || !CF->isDeclaration())
+        continue;
+
+      // Does the function return an integer?
+      if (!CF->getReturnType() || !CF->getReturnType()->isIntegerTy())
+        continue;
+
+      // Is the caller annotated?
+      if (!hasPartialAnnotation(I))
+        continue;
+
+      // Records changes
+      changed = true;
+      LOG("partial-fn", errs() << "Found partial function external call "
+                               << CF->getName().str() << " in "
+                               << F.getName().str() << ".\n");
+
+      // Assumes that the call never happens.
+      builder.SetInsertPoint(&I);
+      CallInst *ci = builder.CreateCall(m_assumeFn, builder.getFalse());
+
+      // Generates a placeholder body.
```

Comment:
do not follow what this code does. please expand comment.

use SmallVector instead of std::vector

Reply:
Revised